scico/loss.py

```diff
@@ -273,17 +283,24 @@ def __init__(
         r"""Initialize a :class:`Loss` object.
 
         Args:
-            y : Measurements
-            A : Forward operator.  Defaults to None.  If None, ``self.A`` is a :class:`.Identity`.
-            scale : Scaling parameter.  Default: 0.5.
-
+            y : Measurement.
+            A : Forward operator. Defaults to None.  If None, ``self.A`` is a :class:`.Identity`.
+            scale : Scaling parameter. Default: 0.5.
         """
         y = ensure_on_device(y)
         super().__init__(y=y, A=A, scale=scale)
 
         #: Constant term in Poisson log likehood; equal to ln(y!)
         self.const: float = gammaln(self.y + 1)  # ln(y!)
 
+        # The "true" Poisson loss is not differentiable at zero, but the
+        # ε that we add to allow evaluation at zero has the side-effect
+        # of making it differentiable there.
+        if isinstance(A, operator.Operator):
+            self.is_smooth = A.is_smooth
+        else:
+            self.is_smooth = None
```

Comment:
@Michael-T-McCann : Any thoughts on this addition? It's a bit ugly mathematically because we're labeling a non-smooth functional as smooth (because we smoothed it at zero), but this seems to be the right choice from a practical/computational perspective.

Reply:
My vote: document the use of epsilon, set `is_smooth=True`.

The only way around the epsilon I come up with is accounting for dark count rate in the forward model.

Reply:
Even with the epsilon it is still not smooth for `Ax < (-epsilon)`. I think a way to get around the mathematical imprecision is to say that the attribute we are looking for is whether it can be used in pgm or not.

If `A(x) = exp(x)` for example, then we can use the current `PoissonLoss` with no problem. But if `A(x)` is linear then we can even get into trouble when using the `epsilon` since `Ax < (-epsilon)` is very likely.

@bwohlberg is the `is_smooth` attribute used for anything other than the check in pgm?

Reply:
> Even with the epsilon it is still not smooth for `Ax < (-epsilon)`. 

Good point. Given that epsilon just shifts the problem, perhaps we should remove it? Alternatively, instead of adding epsilon, set values less than epsilon to epsilon?

> @bwohlberg is the `is_smooth` attribute used for anything other than the check in pgm?

I'm not sure. That may be the only use at the moment. Perhaps the whole `is_smooth` mechanism is worth a re-think. How is the issue of zeros handled in the Poisson loss example with PGM?

Reply:
> Given that epsilon just shifts the problem, perhaps we should remove it? Alternatively, instead of adding epsilon, set values less than epsilon to epsilon?

Yes, this may be a better option. But see below.

> Perhaps the whole is_smooth mechanism is worth a re-think.

Yes. I think if we rejected all non-smooth `L(y, A(x))` for all `x`, we would be unnecessarily impractical. A looser condition of `L(y, A(x))` smooth around `x_0` or in some feasible set could be more practical.

> How is the issue of zeros handled in the Poisson loss example with PGM?

(a) The initial condition is > 0 (almost surely).
(b) There is a non-negative indicator.
(c) Matrix A does not have negative entires
However, it seems like initializing to negatives or zero breaks the example code. I guess this kind of problem is not really that realistic. If we assume `y` to be Poisson, then in what world would `A(x)` be negative for any `x` in the feasible set.

I am leaning towards completely removing the epsilon and then then either `A` or the feasible set needs to be such that `A(x)>0`. (I think that currently would not break anything but I would not bet on it.)

@Michael-T-McCann ?

Reply:
I had assumed the epsilon was necessary for some example or another. If it isn't, sure, remove it.

Reply:
Issue opened in #89